src/buffer.c

```diff
@@ -0,0 +1,98 @@
+#include <stdlib.h>     // malloc, realloc, free
+#include <string.h>     // memset, memmove
+
+#include "buffer.h"
+
+struct zseek_buffer {
+    void *data;
+    size_t size;
+    size_t capacity;
+};
+
+zseek_buffer_t *zseek_buffer_new(size_t capacity)
+{
+    zseek_buffer_t *buffer = malloc(sizeof(*buffer));
+    if (!buffer)
+        goto fail;
+    memset(buffer, 0, sizeof(*buffer));
+
+    if (capacity == 0)
+        return buffer;
+
+    void *data = malloc(capacity);
+    if (!data)
+        goto fail_w_buffer;
+    buffer->data = data;
+    buffer->capacity = capacity;
+
+    return buffer;
+
+fail_w_buffer:
+    free(buffer);
+fail:
+    return NULL;
+}
+
+void zseek_buffer_free(zseek_buffer_t *buffer)
+{
+    if (!buffer)
+        return;
+
+    free(buffer->data);
+    free(buffer);
+}
+
+size_t zseek_buffer_size(zseek_buffer_t *buffer)
+{
+    if (!buffer)
+        return 0;
+
+    return buffer->size;
+}
+
+void *zseek_buffer_data(zseek_buffer_t *buffer)
+{
+    if (!buffer)
+        return NULL;
+
+    if (buffer->size == 0)
+        return NULL;
+
+    return buffer->data;
+}
+
+bool zseek_buffer_push(zseek_buffer_t *buffer, const void *data, size_t len)
+{
+    if (!buffer)
+        return false;
+
+    if (!data)
+        return len == 0;
+
+    size_t new_size = buffer->size + len;
+    if (new_size > buffer->capacity) {
+        // TODO OPT: Use different reallocation strategy? (always power of 2?)
+        size_t new_capacity = 2 * buffer->capacity;
```

Comment:
A good point! I will consider adopting this, since the original reason behind buffering all uncompressed data until the frame is full was that for lz4 < 1.8.0 we compressed the whole frame in one go, but that is now gone.